File: integrations/langgraph/cortadel_langgraph/_runtime.py
"""Reading the ambient LangGraph runtime without letting it throw.

``langgraph.config.get_store()`` is the documented way for a node or tool to reach the store the
graph was compiled with, but it is unforgiving about *where* it is called from — its body is
``get_config()[CONF][CONFIG_KEY_RUNTIME].store`` (``langgraph/config.py``), so it raises:

* ``RuntimeError`` when there is no runnable context at all,
* ``KeyError('__pregel_runtime')`` when there is a runnable context but no LangGraph runtime —
  which is exactly what happens when a tool is invoked standalone, e.g. ``tool.invoke({...})`` in a
  unit test,
* and it returns ``None`` when the graph was compiled without a store.

Memory is an enhancement, never a reason for an agent to crash, so this narrows all three to a
single ``None``.
"""
from __future__ import annotations

from typing import Optional

from langgraph.config import get_store
from langgraph.store.base import BaseStore

__all__ = ["ambient_store"]


def ambient_store() -> Optional[BaseStore]:
    """The store the current graph was compiled with, or ``None`` if there isn't one."""
    try:
        return get_store()
    except (RuntimeError, KeyError, AttributeError):
        return None

File: integrations/langgraph/src/runtime.ts
/**
 * Reading the ambient LangGraph runtime without letting it throw.
 *
 * `getStore()` / `getConfig()` (`@langchain/langgraph`, re-exported from
 * `libs/langgraph-core/src/pregel/utils/config.ts`) are the documented way for a node or tool to
 * reach the store the graph was compiled with. Their JS shape differs from the Python one in two
 * ways that matter here:
 *
 * 1. **`getStore` takes an optional `config`.** Its body is
 *    `(config ?? AsyncLocalStorageProviderSingleton.getRunnableConfig())?.store`, so a node or tool
 *    that was handed a `LangGraphRunnableConfig` can pass it straight in and never touch
 *    AsyncLocalStorage. Python's `get_store()` takes no argument at all. Preferring the explicit
 *    config is why this package threads `config` through every call site.
 * 2. **It throws exactly one error**, a plain `Error` ("Config not retrievable…"), and only when
 *    there is neither an argument nor an ALS config — e.g. a tool invoked standalone in a unit
 *    test, or a web/edge build with no `node:async_hooks`. There is no JS analogue of Python's
 *    `KeyError('__pregel_runtime')`: the store hangs off the config directly rather than off a
 *    runtime object nested inside `configurable`. It returns `undefined` when the graph was
 *    compiled without a store.
 *
 * Memory is an enhancement, never a reason for an agent to crash, so this narrows every one of
 * those outcomes to a single `undefined`.
 */
import { getConfig, getStore, type LangGraphRunnableConfig } from "@langchain/langgraph";
import type { BaseStore } from "@langchain/langgraph-checkpoint";

/** The store the current graph was compiled with, or `undefined` if there isn't one. */
export function ambientStore(config?: LangGraphRunnableConfig): BaseStore | undefined {
  try {
    return getStore(config) ?? undefined;
  } catch {
    return undefined;
  }
}

/**
 * The ambient runnable config, or `undefined` outside a graph.
 *
 * `getConfig()` returns `undefined` rather than throwing when there is no ALS config (Python's
 * `get_config()` raises `RuntimeError` instead), but a web build without `node:async_hooks` can
 * still throw from inside the singleton — hence the guard.
 */
export function ambientConfig(): LangGraphRunnableConfig | undefined {
  try {
    return getConfig() ?? undefined;
  } catch {
    return undefined;
  }
}
